### PR TITLE
Refine SubList and SubListWithProposer

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

Refine `SubList` and `SubListWithProposer`.
- Define common logics to functions: `ConvertHashToSeed`, `SelectRandomCommittee`
- More validation and error handling related to `committeeSize` in `SubListWithProposer` 
- More validation and error handling in `SetSubGroupSize` 

I tried to maintain the semantic of the `SubListWithProposer` except for invalid input handling.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/istanbul/validator/weighted.go [hunk 1]
```diff
@@ -200,6 +199,7 @@ func NewWeightedCouncil(addrs []common.Address, rewards []common.Address, voting
 	if valSet.Size() > 0 {
 		valSet.proposer.Store(valSet.GetByIndex(0))
 	}
+	valSet.SetSubGroupSize(committeeSize)
```
Comment (ehnuje) [inline consensus/istanbul/validator/weighted.go:202]: Why did you change the position of setting the sub group size? Is it related to the change of SetSubGroupSize function?

**Comment by aidan-kwon** [comment]
The last commit allows committee sizes bigger than a validator size since it is more useful to operate Klaytn core cell network.

File: consensus/istanbul/validator/default.go [hunk 2]
```diff
@@ -141,133 +142,63 @@ func (valSet *defaultSet) List() []istanbul.Validator {
...
-		vals := "["
-		for _, v := range valSet.validators {
-			vals += fmt.Sprintf("%s,", v.Address().Hex())
-		}
-		vals += "]"
-		logger.Error("idx should not be negative!", "proposerIdx", proposerIdx, "nextproposerIdx", nextproposerIdx, "proposer", subset[0].Address().Hex(),
-			"nextproposer", subset[1].Address().Hex(), "validators", vals)
+	// return early if the committee size is an edge value
+	if committeeSize == 1 {
+		return []istanbul.Validator{proposer}
```
Comment (winnie-byun) [inline consensus/istanbul/validator/default.go:175]: Why is an empty proposer with only address returned instead of a whole fetched proposer?

Comment (aidan-kwon) [reply consensus/istanbul/validator/default.go:179]: Thank you! I changed the code to use the retuned object.

File: consensus/istanbul/validator/weighted.go (hunk 1 above)
Comment (aidan-kwon) [reply consensus/istanbul/validator/weighted.go:203]: I thought "SubGroupSize" needs to be assigned after the validator size is set.

File: consensus/istanbul/validator/default.go [hunk 3]
```diff
@@ -131,7 +127,18 @@ func (valSet *defaultSet) SubGroupSize() uint64 {
 	return valSet.subSize
 }
 
+// SetSubGroupSize sets committee size of the valSet.
 func (valSet *defaultSet) SetSubGroupSize(size uint64) {
+	if size == 0 {
```
Comment (KimKyungup) [inline consensus/istanbul/validator/default.go:132]: This condition can be checked in next `if statement`.
We can skip this.

Comment (KimKyungup) [reply consensus/istanbul/validator/default.go:132]: Ignore this. kaka

File: consensus/istanbul/validator/validator.go [hunk 4]
```diff
@@ -63,3 +67,75 @@ func ExtractValidators(extraData []byte) []common.Address {
...
+func SelectRandomCommittee(validators []istanbul.Validator, committeeSize uint64, seed int64, proposerIdx int, nextProposerIdx int) []istanbul.Validator {
+	// ensure validator indexes are valid
+	if proposerIdx < 0 || nextProposerIdx < 0 || proposerIdx == nextProposerIdx {
+		logger.Error("invalid indexes of validators", "proposerIdx", proposerIdx, "nextProposerIdx", nextProposerIdx)
+		return nil
+	}
+
+	// ensure committeeSize and proposer indexes are valid
+	validatorSize := len(validators)
+	if validatorSize < int(committeeSize) || validatorSize < proposerIdx || validatorSize < nextProposerIdx {
```
Comment (KimKyungup) [inline consensus/istanbul/validator/validator.go:99]: This condition `validatorSize < int(committeeSize)` returning nil is ok?

File: consensus/istanbul/validator/weighted.go [hunk 5]
```diff
@@ -290,97 +296,82 @@ func (valSet *weightedCouncil) List() []istanbul.Validator {
...
+	// return early if the committee size is an edge value
+	if committeeSize == 1 {
+		return []istanbul.Validator{proposer}
 	}
-
-	if proposerIdx == nextproposerIdx {
-		logger.Error("fail to make propser", "current proposer idx", proposerIdx, "next idx", nextproposerIdx)
+	if committeeSize >= validatorSize {
+		return validators
 	}
```
Comment (laggu) [inline consensus/istanbul/validator/weighted.go:334]: how about moving these codes to top of the function.
we can just return if the commiteeSize is bigger than validatorSize

File: consensus/istanbul/validator/weighted.go [hunk 6]
```diff
@@ -290,97 +296,82 @@ func (valSet *weightedCouncil) List() []istanbul.Validator {
...
+			"addr", nextProposer.Address().String(), "index", nextProposerIdx)
+		return validators
 	}
 
-	for i := uint64(0); i < valSet.subSize-2; i++ {
-		committee[i+2] = valSet.validators[indexs[i]]
+	// seed will be used to select a random committee
+	seed, err := ConvertHashToSeed(prevHash)
+	if err != nil {
+		logger.Error("failed to covert hash to seed", "prevHash", prevHash, "err", err)
```
Comment (laggu) [inline consensus/istanbul/validator/weighted.go:362]:
```suggestion
		logger.Error("failed to convert hash to seed", "prevHash", prevHash, "err", err)
```

Comment (aidan-kwon) [reply consensus/istanbul/validator/weighted.go]: Thanks!

File: consensus/istanbul/validator/weighted.go (hunk 5 above)
Comment (aidan-kwon) [reply consensus/istanbul/validator/weighted.go]: Thank you. I moved it!

**Comment by aidan-kwon** [comment]
Reviewers, please take a look.

File: consensus/istanbul/validator/validator.go [hunk 7]
```diff
@@ -63,3 +67,75 @@ func ExtractValidators(extraData []byte) []common.Address {
...
+		logger.Error("fail to make sub-list of validators", "hash", hash.Hex(), "seed", seed, "err", err)
+		return 0, err
+	}
+	return seed, nil
+}
+
+// SelectRandomCommittee composes a committee selecting validators randomly based on the seed value.
+// It returns nil if the given committeeSize is bigger than validatorSize or proposer indexes are invalid.
+func SelectRandomCommittee(validators []istanbul.Validator, committeeSize uint64, seed int64, proposerIdx int, nextProposerIdx int) []istanbul.Validator {
+	// ensure validator indexes are valid
```
Comment (KimKyungup) [inline consensus/istanbul/validator/validator.go:91]:
```suggestion
	// ensure proposers indexes are valid
```

File: consensus/istanbul/validator/weighted.go [hunk 8]
```diff
@@ -280,7 +281,12 @@ func (valSet *weightedCouncil) SubGroupSize() uint64 {
 	return valSet.subSize
 }
 
+// SetSubGroupSize sets committee size of the valSet.
 func (valSet *weightedCouncil) SetSubGroupSize(size uint64) {
+	if size == 0 {
+		logger.Error("cannot assign committee size to 0")
```
Comment (KimKyungup) [inline consensus/istanbul/validator/weighted.go:287]: Crit? 
```suggestion
		logger.Crit("cannot assign committee size to 0")
```

File: consensus/istanbul/validator/validator.go (hunk 7 above)
Comment (aidan-kwon) [reply consensus/istanbul/validator/validator.go:91]: I think "validator" is good to go because the index is indexes of `validators` variable

File: consensus/istanbul/validator/weighted.go (hunk 8 above)
Comment (aidan-kwon) [reply consensus/istanbul/validator/weighted.go:287]: I think `error` is enough. It doesn't change `valSet.subSize` value when `size` is 0.

**Comment by KimKyungup** [comment]
@ehnuje PTAL. :)
